oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/Optimizer.kt

```diff
@@ -0,0 +1,358 @@
+                    is SimResult.Success -> {
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation Succeed: Result [${result.result}]"))
+                        viewData.addResult(Result(result.name, "Success", inputVector.toString(), result.result.toString()))
+                    }
+                    is SimResult.Failure -> {
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation Failed: Due to\n${result.exception}"))
+                        viewData.addResult(Result(result.name, "Error", inputVector.toString(), "Failure: \n${result.exception}"))
+                    }
+                    is SimResult.TimeOut -> {
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation Canceled to timed out after ${sim.timeOut!!.toMillis()} millisecond"))
+                        viewData.addResult(Result(result.name, "Canceled", inputVector.toString(), "Timedout"))
+                    }
+                    is SimResult.TimeOutFailure -> {
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation try to cancel Timed out after ${sim.timeOut!!.toMillis()} millisecond but Failed: Due to\n${result.exception}"))
+                        viewData.addResult(Result(result.name, "Cancel Failed", inputVector.toString(), "Cancellation failed:\n${result.exception}"))
+                    }
+                }
+                i++
+            }
+        }
+        viewData.updateStatusMessage("Idle")
+    }
+
+
+    fun stopOptimization() {
+        state = State.Idle
+    }
+
+    fun syncAll() = GlobalScope.launch {
+        for (simName in simulationsByName.keys) {
+            syncConfigFor(simName)
+        }
+    }
+
+
+    fun cancelAll() = GlobalScope.launch {
+        for ((name, sim) in simulationsByName) {
+            val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+            ).build()
+
+            sim.input.onNext(message)
+        }
+    }
+
+    fun cancel(name: String) = GlobalScope.launch {
+        val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+        ).build()
+
+        simulationsByName.getValue(name).input.onNext(message)
+    }
+
+    fun setDuration(nodeName: String?, timeOut: Duration?) {
+        if (nodeName == null || !simulationsByName.containsKey(nodeName)) {
+            return
+        }
+        val node: Simulation = simulationsByName.getValue(nodeName)
+        val newNode = node.copy(
+                timeOut = timeOut
+        )
+        simulationsByName += nodeName to newNode
+    }
+
+    fun disconnectAll() = GlobalScope.launch {
+        for ((name, sim) in simulationsByName) {
+            try {
+                sim.input.onCompleted()
```

Comment:
so, what do we do to actually stop the coroutine thats producing data for that pipe? shouldnt we shut it down too?

Reply:
I am not exactly sure what I was doing here. My decision on these are dropping these until we know what we want to do with these.